libs/@local/harpc/wire-protocol/src/codec/buffer.rs

```diff
@@ -0,0 +1,347 @@
+use std::io::Cursor;
+
+use bytes::{Buf, BufMut, Bytes};
+use error_stack::{Report, Result};
+
+pub(crate) trait Number {
+    const WIDTH: usize;
+
+    fn unchecked_read_from_buf<B>(buf: B) -> Self
+    where
+        B: Buf;
+
+    fn unchecked_write_to_buf<B>(self, buf: B)
+    where
+        B: BufMut;
+}
+
+impl Number for u8 {
+    const WIDTH: usize = 1;
+
+    fn unchecked_read_from_buf<B>(mut buf: B) -> Self
+    where
+        B: Buf,
+    {
+        buf.get_u8()
+    }
+
+    fn unchecked_write_to_buf<B>(self, mut buf: B)
+    where
+        B: BufMut,
+    {
+        buf.put_u8(self);
+    }
+}
+
+impl Number for u16 {
+    const WIDTH: usize = 2;
+
+    fn unchecked_read_from_buf<B>(mut buf: B) -> Self
+    where
+        B: Buf,
+    {
+        buf.get_u16()
+    }
+
+    fn unchecked_write_to_buf<B>(self, mut buf: B)
+    where
+        B: BufMut,
+    {
+        buf.put_u16(self);
+    }
+}
+
+impl Number for u32 {
+    const WIDTH: usize = 4;
+
+    fn unchecked_read_from_buf<B>(mut buf: B) -> Self
+    where
+        B: Buf,
+    {
+        buf.get_u32()
+    }
+
+    fn unchecked_write_to_buf<B>(self, mut buf: B)
+    where
+        B: BufMut,
+    {
+        buf.put_u32(self);
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, thiserror::Error)]
+pub enum BufferError {
+    #[error("early end of stream")]
+    EarlyEndOfStream,
+    #[error("not enough capacity to write to buffer")]
+    NotEnoughCapacity,
+}
+
+pub struct Buffer<'a, B>(&'a mut B);
+
+impl<'a, B> Buffer<'a, B> {
+    pub fn new(buffer: &'a mut B) -> Self {
+        Self(buffer)
+    }
+}
+
+impl<'a, B> Buffer<'a, B>
+where
+    B: Buf,
+{
+    pub(crate) fn next_number<N: Number>(&mut self) -> Result<N, BufferError> {
+        if self.0.remaining() < N::WIDTH {
+            return Err(Report::new(BufferError::EarlyEndOfStream));
+        }
+
+        Ok(N::unchecked_read_from_buf(&mut self.0))
+    }
+
+    pub(crate) fn next_bytes(&mut self, at: usize) -> Result<Bytes, BufferError> {
+        if self.0.remaining() < at {
+            return Err(Report::new(BufferError::EarlyEndOfStream));
+        }
+
+        Ok(self.0.copy_to_bytes(at))
+    }
+
+    pub(crate) fn next_array<const N: usize>(&mut self) -> Result<[u8; N], BufferError> {
+        if self.0.remaining() < N {
+            return Err(Report::new(BufferError::EarlyEndOfStream));
+        }
+
+        let mut bytes = [0; N];
+        self.0.copy_to_slice(&mut bytes);
+
+        Ok(bytes)
+    }
+
+    pub(crate) fn next_discard(&mut self, count: usize) -> Result<(), BufferError> {
```

Comment:
As we don't return anything I think I would rename it to `discard`

Reply:
addressed in [05d2e73](https://github.com/hashintel/hash/pull/4389/commits/05d2e73b18cbc0207217af53062812175c631af9)